Add KmsKeyId to LogGroup

ikben opened the pull request:
~This is a Work in Progress, if that's better as an issue, let me know.~

I tried to tackle the difficult parts first, which for me besides the whole of Java is writing test. I haven't started writing code yet. Assistance with both parts is very welcome.

I'm opening this PR mainly to get feedback on the approach. Currently I have the following questions related to the work already in here:

- Spec is step 1, so that can already get a decent review :)
- Is this the way to update the tests, or do we prefer completely new tests that only deal with the KmsKeyId

I also have question for the work that isn't in here yet:

- There are three related methods in the API: "CreateLogGroup" with the "kmsKeyId" parameter, "AssociateKmsKey" and "DisassociateKmsKey"
- This allows to approaches in the CreateHandler. Doing it in two calls (CreateLogGroup without kmsKeyId and AssociateKmsKey if it's not Null on the model), which leaves the LogGroup without encryption for some time (and in that time some messages might appear). Or doing it in one call, which uses a different flow for the createHandler and the updateHandler.
- In the updateHandler we have to do it in two steps, which is something that we either shouldn't support (only allowing going from encrypted to plaintext and vice versa) or should highlight in the documentation. The risk of log lines arriving while this change is in progress is even higher than with the CreateHandler.

We should probably decide on the best approach before writing the code.


-----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

rjlohan commented:
Not super important, but you can do 'Draft' PRs to signal WIP. We can review them entirely, but they're not merge-able.

rjlohan commented on aws-logs-loggroup/aws-logs-loggroup.json:
```diff
@@ -3,8 +3,14 @@
     "description": "Resource schema for AWS::Logs::LogGroup",
     "sourceUrl": "https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs.git",
     "properties": {
+        "KmsKeyId": {
```
Will need to update the `handlers` section too with additional IAM keys here based on the API decision.

ikben replied:
I'm assuming you will need "logs:AssociateKmsKey" if you do it in one action too. It's not yet in the docs (https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/permissions-reference-cwl.html)

ikben replied:
On that note. Why is `"logs:PutRetentionPolicy"` a permission for the UpdateHandler, but not for the CreateHandler? They both do that in a separate call ([see here](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs/blob/c72bc6d300a1ba3db06862eb22e7f81731fc9db0/aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/CreateHandler.java#L50)):

rjlohan replied:
Unclear what permissions are needed here. Sometimes the explicit API permission isn't checked when wrapped in another API call.

On the second point, that's totally a miss. This type wouldn't work when run in Private mode.

ikben replied:
You were right, `logs:CreateLogGroup` is all you need

rjlohan commented on aws-logs-loggroup/aws-logs-loggroup.json:
```diff
@@ -3,8 +3,14 @@
+            "description": "The Amazon Resource Name (ARN) of the CMK to use when encrypting log data.",
+            "type": "string",
+            "maxLength": 256,
+            "pattern": "^arn:[a-z0-9-]+:kms:[a-z0-9-]+:\\d{12}:(key|alias)/.+\\Z"
```
Will want to come back to this so we can $ref an `AWS::KMS::Key` resource's `Arn` property when that's available.

ikben commented:
Current blocked from (easily/comfortably) writing the implementation by #25  (and a little bit by #28, but I got that working locally by pulling in #29 )

rjlohan commented:
If you merge from master, you'll get a Travis build for this PR from now on. 👍

rjlohan commented:
@ikben how's this PR going?

ikben commented:
I didn't get to pulling in the recent changes yet, hopefully I will have some time to look at it in the next few days

ikben commented:
I think this is what the implementation would look like. Please note:

- contract tests still fail, because of #25 additionally they now also do not return the KmsKeyId (for the same reason). The ARN is the bigger issue in my opinion, until there is a decision made about haw to tackle that, it does not make a lot of sense to fix that here.
- 7596883d220658aa5f78113e94339a289f3340e4 and ff7dc2d813ab662e5d99abd04be9c60cdd18ebce are useful even without adding the KmsKeyId, so if it turns out that there is still a lot more work on this, they can be cheryy-picked.

rjlohan commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:
```diff
@@ -28,6 +30,15 @@
 
+        // It can take up to five minutes for the (dis)associate operation to take effect
+        // It's unclear from the documentation if that state can be checked via the API.
+        // https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html
```
Are we emitting ProgressEvents with `IN_PROGRESS` to check on the status or it seems like that's not actually discoverable?

wbingli replied:
Where does the 5 mins come from?  As far as I see there is no such status in the log group. It may take some time for data plane to take effect, but from control plane perspective, it's done once API call successful.

ikben replied:
The docs say "After you associate or disassociate a CMK from a log group, it can take up to five minutes for the operation to take effect." To me that sounds like the data plane needing some time, like Wenbing described,and what I assumed when writing the code. There is no separate API call to check on the status and it's hard to  test because "it is immediately  present when I do a read call", still falls into the "up to 5 minutes" bucket.

ikben replied:
The docs say "After you associate or disassociate a CMK from a log group, it can take up to five minutes for the operation to take effect." To me that sounds like the data plane needing some time, like Wenbing described,and what I assumed when writing the code. There is no separate API call to check on the status and it's hard to test because "it is immediately present when I do a read call", still falls into the "up to 5 minutes" bucket.

miparnisari commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:
```diff
@@ -69,6 +80,44 @@ private void putRetentionPolicy(final AmazonWebServicesClientProxy proxy,
+        try {
+            proxy.injectCredentialsAndInvokeV2(disassociateKmsKeyRequest,
+                    ClientBuilder.getClient()::disassociateKmsKey);
+        } catch (final ResourceNotFoundException e) {
```
[There are more exceptions that can be thrown](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_DisassociateKmsKey.html#API_DisassociateKmsKey_Errors)

ikben replied:
I think that it makes sense to look at the for the whole resource (other calls don't catch everything either).

Until there is a documentation page with the differences between all exceptions (I might be overlooking that - a quick google search gave me nothing), I'd probably cast the to a InternalFailure anyway

miparnisari replied:
I think it's important to throw proper exceptions because some may be retriable and some may not be. https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-contract-errors.html

ikben replied:
I agree. I'll catch them for my changes, and I'll add a follow-up issue to add the same thing to existing code

miparnisari commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:
```diff
@@ -69,6 +80,44 @@ private void putRetentionPolicy(final AmazonWebServicesClientProxy proxy,
+        try {
+            proxy.injectCredentialsAndInvokeV2(associateKmsKeyRequest,
+                    ClientBuilder.getClient()::associateKmsKey);
+        } catch (final ResourceNotFoundException e) {
```
[There are more exceptions that can be thrown](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_AssociateKmsKey.html#API_AssociateKmsKey_Errors)

PatMyron commented on aws-logs-loggroup/README.md:
```diff
@@ -15,3 +15,20 @@ automatically overwritten.
+
+## Running Contract Tests
+
+1. Create a KMS CMK for use with CloudWatch Logs (see [the CloudWatch
```
Being the infrastructure-as-code service, we should probably just get this step as code 😛

PatMyron commented on aws-logs-loggroup/README.md:
```diff
@@ -15,3 +15,20 @@ automatically overwritten.
+
+1. Create a KMS CMK for use with CloudWatch Logs (see [the CloudWatch
+    Documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html)).
+2. Edit `overrides.json` so that every `/KmsKeyId` uses the ARN of that key.
+3. Package the code with Maven (`mvn package`)
+4. Run `sam local start-lambda` in one terminal
+5. Run `cfn test` in another terminal. CFN test will use your credentials to test
+    the resource handlers in your account.
```
thoughts on [not explicitly ordering markdown lists](https://guides.github.com/features/mastering-markdown/#syntax) and making these commands [more copy-pasteable](https://github.com/PatMyron/aws-cloudformation-resource-providers-logs/blob/kms-key-id/aws-logs-loggroup/README.md)?

   ```shell
   echo '{
     "CREATE": {
       "/KmsKeyId": "$KEY_ARN"
     }
   }' > overrides.json # so that every /KmsKeyId uses the ARN of that key.
   mvn package # package the code with Maven
   sam local start-lambda # in one terminal
   cfn test # in another terminal. CFN test will use your credentials to test the resource handlers in your account
   ```

ikben replied:
I disagree with GitHub there, the main goal of markdown is to be readable as a text file :)

ikben commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Translator.java:
```diff
@@ -62,19 +78,25 @@ static ResourceModel translateForRead(final DescribeLogGroupsResponse response)
                 .arn(logGroupArn)
                 .logGroupName(logGroupName)
                 .retentionInDays(retentionInDays)
+                .kmsKeyId(kmsKeyId)
```
True, I followed the code above (original like 69-73), that also breaks that rule

ikben replied:
Actually, I might need some java help here. I'd assume the rpdk would take care of that, since all the properties are properties of the ResourceModel class

miparnisari replied:
Hmm. I don't think the rpdk is checking this. @rjlohan do you have any recommendations here? Other than checking for `null` before building the model's properties?

ikben replied:
@rjlohan Do you have any feedback on this?

rjlohan replied:
The real reason for this is to disambiguate these scenarios;
* property not set/present
* property explicitly set to null/empty

It's not a strict requirement, hence it's a `SHOULD NOT` rather than a `MUST NOT`. I'm ok to proceed as is.

ikben commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:
```diff
@@ -19,7 +21,7 @@
+        // Everything except RetentionPolicyInDays KmsKeyId is createOnly
         final ResourceModel model = request.getDesiredResourceState();
 
         if (model.getRetentionInDays() == null) {
```
Is this optimisation needed? The delete will still succeed if there is no retention in the current state.

ikben commented on aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:
```diff
@@ -28,6 +30,15 @@
+        // It can take up to five minutes for the (dis)associate operation to take effect
+        // It's unclear from the documentation if that state can be checked via the API.
+        // https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html
+        if (model.getKmsKeyId() == null) {
```
(see reply to the similar line of code for RetentionInDays)

miparnisari commented on aws-logs-loggroup/README.md:
````diff
@@ -15,3 +15,44 @@ automatically overwritten.
+You can execute the following commands to run the tests.
+You will need to have docker installed and running.
+
+```shell script
````
```suggestion
```bash
```

ikben commented:
I had rewrote a few if statements and added a bunch of tests to get the required coverage.\

The issue about having model properties that can be `null` is still open, but I'll wait on feedback for the best approach there

OfficialTermi commented:
Thanks for the implementation! We've just tested this out and it seems not to be in production as of yet.

A bit of topic, but important nonethelss: Is there a way to get notified or keep track of changes like these as they make their way through quality assurance and to production? As far as I'm aware, the only way right now is trial and error, which is quite frustrating.

ikben commented:
@OfficialTermi 

- The best way to see that should be the [CloudFormation Coverage Roadmap Issue](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/128). Unfortunately, that issue hasn't been moved to the roadmap yet (for the AWS-folks here, is the coverage roadmap updated if the PR is coming from outside of AWS?)
- That does not mean you have to do it with trial and error though.  You can look at:
    - [The CloudFormation Release History](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/ReleaseHistory.html) (it also has an RSS feed)
    - [The CloudFormation Resource Specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html)
    - The API, using the CLI to describe the type definition and check the Schema for "KmsKeyId": `aws cloudformation describe-type --type RESOURCE --type-name AWS::Logs::LogGroup | grep KmsKeyId`

miparnisari commented on aws-logs-loggroup/dev-resources.yaml:
```diff
@@ -0,0 +1,39 @@
+  CmkAlias:
+    Type: "AWS::KMS::Alias"
+    Properties:
+      AliasName: !Sub "arn:${AWS::Partition}:kms:${AWS::Region}:${AWS::AccountId}:alias/${AWS::StackName}"
```
@ikben this is giving me 

```
"ResourceStatus": "CREATE_FAILED",
"ResourceStatusReason": "Model validation failed (#/AliasName: failed validation constraint for keyword [pattern])"
```

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kms-alias.html#aws-resource-kms-alias-properties

ikben replied:
@miparnisari Thanks for the fix. Yours does follow what's in the documentation.

At the time of writing, it did deploy though (I checked the deleted stack in my account, it has the long !Sub). So make sure that it's not the StackName that is causing the failure.
